src/main/scala/viper/gobra/util/ViperChopper.scala

```diff
@@ -0,0 +1,644 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) 2011-2020 ETH Zurich.
+
+package viper.gobra.util
+
+
+import viper.silver.{ast => vpr}
+import viper.gobra.util.ViperChopper.Vertex.Always
+
+import scala.annotation.tailrec
+import scala.collection.{immutable, mutable}
+
+object ViperChopper {
+
+  /**
+    * chops 'choppee' into multiple Viper programs.
+    * @param choppee Targeted program.
+    * @param isolate Specifies which members of the program should be verified.
+    *                If none, then all members that induce a proof obligation are verified.
+    * @param bound Specifies upper bound on the number of returned programs.
+    *              If none, then maximum number of programs is returned.
+    * @param penalty Specifies penalty of merging programs.
+    * @return Chopped programs.
+    */
+  def chop(choppee: vpr.Program)(
+    isolate: Option[vpr.Member => Boolean] = None,
+    bound: Option[Int] = Some(1),
+    penalty: Penalty[Vertex] = Penalty.Default
+  ): Vector[vpr.Program] = {
+
+    val (n, isolatesIds, edges, idToVertex, inverse) = ViperGraph.toGraph(choppee, isolate)
+    val programs = Cut.boundedCut(n, isolatesIds, edges, idToVertex)(bound, penalty)
+    programs flatMap (list => inverse(list.toSet))
+  }
+
+  object Cut {
+
+    /**
+      * Returns a set of chopped programs where the number of programs is bounded by `bound`.
+      * @param N Number of nodes.
+      * @param nodes Nodes that must be included.
+      * @param edges Edges of the dependency graph.
+      * @param idToVertex Mapping from node ids to vertices used for price computations.
```

Comment:
why do you operate on `Int` and use a mapping function to retrieve the corresponding `Vertex`? Is this better for performance?

Reply:
Using the ints, I can use arrays instead of maps everywhere.

Reply:
I tested different approaches, where I varried the information inside the vertices and the representation of the nodes. The current one seemed to me to have the best performance in terms of transforming to the node representation and then executing the algorithms on it.

Reply:
Thanks